### PR TITLE
Refactor bokeh.server module

### LINKED ISSUE #6439: Server(..., io_loop=io_loop, num_procs=0) --> RuntimeError
From the [flask_embed](https://github.com/bokeh/bokeh/blob/master/examples/howto/server_embed/flask_embed.py) example:

```
bokeh_app = Application(FunctionHandler(modify_doc))

io_loop = IOLoop.current()

server = Server({'/bkapp': bokeh_app}, io_loop=io_loop, allow_websocket_origin=["localhost:8080"])
server.start()
```

`Server.__init__` takes a [kwarg](https://github.com/bokeh/bokeh/blob/master/bokeh/server/server.py#L81) `num_procs` that I'd like to set to 0 to fork a reasonable number of processes*.  But since the `IOLoop` has already been initialized, that raises:

```
RuntimeError: Cannot run in multiple processes: IOLoop instance has already been initialized. You cannot call IOLoop.instance() before calling start_processes()
```

Most of the examples I've seen calling `Server.__init__` pass an `IOLoop` object.  So is there another place `num_procs` is supposed to be passed?

\* As it is, only one request is getting served at a time.  That suggests that I'm somehow using Tornado wrong, but that's a question for StackExchange.

python2 .7.12 running on AWS linux.
bokeh 0.12.5

### PR DESCRIPTION
**Pull request by bryevdv**
issues: closes #6439

* All protocol-related code was moved to a new module ``bokeh.protocol``

* The ``Server`` class was split into ``Server`` (no API changes) and a new ``BaseServer`` that can be used when explicit coordination of low level Tornado components is required.

* Missing sections added to reference guide, and docsctring coverage greatly expanded.

* Old cruft code to handle Tornado < 4.3 (which is no longer supported) was removed.

ping @mattpap @canavandl I still need to update a few examples and hopefully add a few more tests, but this is generally ready. Apologies for some extraneous updates included but refactoring the docs code can be a pain and sometimes I need to get similar kinds of changes out of the way in the same place/time for my own sanity.   

The majority of this PR is adding documentation and some file moves. The files to look for for code changes are mostly `server.py` and `tornado.py`

### CODE REVIEW

File: bokeh/sphinxext/_templates/options_detail.rst [hunk 1]
```diff
@@ -1,11 +1,6 @@
...
--------
+Keyword Args:
 
 {% for opt in opts %}
-{{ opt['name'] }}
-    *type*: {{ opt['type'] }}
-    {% if opt['default'] %} default: {{ opt['default']}} {% endif %}
-
-    {{ opt['doc']|indent(4) }}
+    {{ opt['name'] }} ({{ opt['type'] }}) : {{ opt['doc']|indent(8) }} (default: {{ opt['default']}})
```
Comment (bryevdv) [inline bokeh/sphinxext/_templates/options_detail.rst]: This makes options arguments be formatted using standard google-style docstring ReST format

File: bokeh/sphinxext/bokeh_options.py [hunk 2]
```diff
@@ -81,7 +81,7 @@ def run(self):
             opts.append(dict(
                 name=prop_name,
                 type=descriptor.property._sphinx_type(),
-                default=None,
+                default=repr(descriptor.instance_default(options_obj)),
```
Comment (bryevdv) [inline bokeh/sphinxext/bokeh_options.py]: This makes sure the default value shows up when options are documented.

File: bokeh/server/util.py [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+
+    if len(pattern) > len(host):
+        return False
+
+    for h, p in zip(host, pattern):
+        if h == p or p == '*':
+            continue
+        else:
+            return False
+    return True
```
Comment (bryevdv) [inline bokeh/server/util.py]: These functions were scattered in other files. Collected them here and added tests.

File: bokeh/server/tornado.py [hunk 4]
```diff
@@ -158,11 +151,18 @@ def __init__(self, applications,
...
             raise ValueError("stats_log_frequency_milliseconds must be > 0")
 
-        self._websocket_origins = set(extra_websocket_origins)
+        self._keep_alive_milliseconds = keep_alive_milliseconds
+        self._check_unused_sessions_milliseconds = check_unused_sessions_milliseconds
+        self._unused_session_lifetime_milliseconds = unused_session_lifetime_milliseconds
+        self._stats_log_frequency_milliseconds = stats_log_frequency_milliseconds
+
+        if extra_websocket_origins is None:
+            self._websocket_origins = set()
```
Comment (bryevdv) [inline bokeh/server/tornado.py]: Note to self: this branch needs a test

File: bokeh/server/tests/test_server.py [hunk 5]
```diff
@@ -619,7 +586,7 @@ def test__ioloop_not_forcibly_stopped():
     application = Application()
     loop = IOLoop()
     loop.make_current()
-    server = Server(application, ioloop=loop)
+    server = Server(application, io_loop=loop)
```
Comment (bryevdv) [inline bokeh/server/tests/test_server.py]: bug in the test previously undiscovered due to loose kwargs handling

File: bokeh/server/tests/test_server.py [hunk 6]
```diff
@@ -25,41 +25,6 @@
...
-def test__create_hosts_whitelist_bad_host_raises():
-    with pytest.raises(ValueError):
-        server._create_hosts_whitelist([""], 1000)
-
-    with pytest.raises(ValueError):
-        server._create_hosts_whitelist(["a:b:c"], 1000)
-
-    with pytest.raises(ValueError):
-        server._create_hosts_whitelist([":80"], 1000)
-
```
Comment (bryevdv) [inline bokeh/server/tests/test_server.py]: These were moved to `test_utils`

File: bokeh/server/tests/test_tornado.py [hunk 7]
```diff
@@ -13,41 +13,6 @@
...
-    assert False == tornado.check_whitelist("192.168.0.1", ['192.168.0.*:5006'])
-    assert True == tornado.check_whitelist("192.168.0.1", ['192.168.0.*:80'])
-    assert True == tornado.check_whitelist("foobarbaz", ['*'])
-    assert True == tornado.check_whitelist("foobarbaz", ['*:*'])
-    assert True == tornado.check_whitelist("foobarbaz", ['*:80'])
-    assert False == tornado.check_whitelist("foobarbaz", ['*:5006'])
-    assert True == tornado.check_whitelist("foobarbaz:5006", ['*'])
-    assert True == tornado.check_whitelist("foobarbaz:5006", ['*:*'])
-    assert True == tornado.check_whitelist("foobarbaz:5006", ['*:5006'])
-
```
Comment (bryevdv) [inline bokeh/server/tests/test_tornado.py]: moved to `test_util`

File: bokeh/server/task.py [hunk 8]
```diff
@@ -1,11 +0,0 @@
...
-
-'''
-from __future__ import absolute_import
-
-# TODO (havocp) this class may be unused
-class ServerTask(object):
-    '''
-
-    '''
-    pass
```
Comment (bryevdv) [inline bokeh/server/task.py]: This was cruft that was never fully realized/implemented, removed

File: bokeh/server/tornado.py [hunk 9]
```diff
@@ -357,15 +416,6 @@ def log_stats(self):
...
             c.send_ping()
-
-    @gen.coroutine
-    def run_in_background(self, _func, *args, **kwargs):
-        """
-        Run a synchronous function in the background without disrupting
-        the main thread. Useful for long-running jobs.
-        """
-        res = yield self._executor.submit(_func, *args, **kwargs)
-        raise gen.Return(res)
```
Comment (bryevdv) [inline bokeh/server/tornado.py]: This was cruft that was never fully realized/implemented or used. Its not used by our code and it's not mentioned in any docs or issues, removing without comment.

File: bokeh/server/tornado.py [hunk 10]
```diff
@@ -16,134 +14,129 @@
...
                  redirect_root=True):
 
+        # This will be set when initialize is called
+        self._loop = None
+
+        if isinstance(applications, Application):
+            applications = { '/' : applications }
+
+        if prefix is None:
+            prefix = ""
```
Comment (bryevdv) [inline bokeh/server/tornado.py]: This branch also probably needs a test

**Comment by bryevdv** [comment]
@mattpap @canavandl I've left some comments to help navigate the large changeset

**Comment by bryevdv** [comment]
I'm not sure what's going on with the py27 test failure:
```
+bokeh info
Traceback (most recent call last):
  File "/home/travis/miniconda/bin/bokeh", line 5, in <module>
    from bokeh.__main__ import main
  File "/home/travis/miniconda/lib/python2.7/site-packages/bokeh/__main__.py", line 3, in <module>
    from bokeh.command.bootstrap import main as _main
  File "/home/travis/miniconda/lib/python2.7/site-packages/bokeh/command/__init__.py", line 18, in <module>
    _build_docstring()
  File "/home/travis/miniconda/lib/python2.7/site-packages/bokeh/command/__init__.py", line 10, in _build_docstring
    from . import subcommands
  File "/home/travis/miniconda/lib/python2.7/site-packages/bokeh/command/subcommands/__init__.py", line 31, in <module>
    all = _collect()
  File "/home/travis/miniconda/lib/python2.7/site-packages/bokeh/command/subcommands/__init__.py", line 19, in _collect
    mod = import_module("." + modname, __package__)
  File "/home/travis/miniconda/lib/python2.7/importlib/__init__.py", line 37, in import_module
    __import__(name)
  File "/home/travis/miniconda/lib/python2.7/site-packages/bokeh/command/subcommands/serve.py", line 293, in <module>
    from bokeh.server.server import Server
  File "/home/travis/miniconda/lib/python2.7/site-packages/bokeh/server/server.py", line 33, in <module>
    from .util import bind_sockets, create_hosts_whitelist
  File "/home/travis/miniconda/lib/python2.7/site-packages/bokeh/server/util.py", line 8, in <module>
    from tornado import netutil
ImportError: cannot import name netutil
```

The failing import `from tornado import netutil` was previously at the top level of `server.py` (unconditional). I sanity checked a conda env with py2.7 and tornado 4.5.1 and the import works fine.

**Comment by bryevdv** [comment]
OK not sure what weird interaction is happening to affect the CI but I've protected the import in `serve.py`. This should allow `bokeh info` to function even if tornado is not installed, which is a good thing in any case since some folks do use that configuration.

**Comment by bryevdv** [comment]
OK, well now it just fails in actual tests:
```
______________ ERROR collecting bokeh/server/tests/test_server.py ______________
ImportError while importing test module '/home/travis/build/bokeh/bokeh/bokeh/server/tests/test_server.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
bokeh/server/tests/test_server.py:15: in <module>
    import bokeh.server.server as server
bokeh/server/server.py:33: in <module>
    from .util import bind_sockets, create_hosts_whitelist
bokeh/server/util.py:8: in <module>
    from tornado import netutil
E   ImportError: cannot import name netutil
```
I am at a serious loss here, this is the *exact same* import that was unprotected in `server.py` only now it's in a sub module that `server.py` imports

**Comment by bryevdv** [comment]
`util.py` was missing `from __future__ import absolute_import` and there was a `tornado.py` in the same dir. We should possible add a code quality check that enforces that every python module imports this at the top

File: bokeh/protocol/message.py [hunk 11]
```diff
@@ -1,5 +1,41 @@
 ''' Provide a base class for all Bokeh Server Protocol message types.
 
+Boker messages are comprised of a sequence of JSON fragments. Specified as
```
Comment (canavandl) [inline bokeh/protocol/message.py]: Bokeh or broker?

Comment (bryevdv) [inline bokeh/protocol/message.py]: I'll fix in general 0127 docs pr

**Pull request merged**